addons/website/models/website_visitor.py

```diff
@@ -4,139 +4,203 @@
-        if website_page.is_tracked:
-            if not visitor_sudo:
-                # If visitor does not exist
-                visitor_sudo = self._create_visitor(website_page.id)
-                sign = visitor_sudo._get_visitor_sign().get(visitor_sudo.id)
-                expiration_date = datetime.now() + timedelta(days=100*365)  # never expire
-                response.set_cookie('visitor_id', sign, expires=expiration_date.timestamp())
-            else:
-                # Add page even if already in visitor_page_ids as checks on relations are done in many2many write method
-                vals = {
-                    'last_connection_datetime': datetime.now(),
-                    'visitor_page_ids': [(0, 0, {'page_id': website_page.id, 'visit_datetime': datetime.now()})],
-                }
-                if visitor_sudo.last_connection_datetime < (datetime.now() - timedelta(hours=8)):
-                    vals['visit_count'] = visitor_sudo.visit_count + 1
-                if not visitor_sudo.active:
-                    vals['active'] = True
-                visitor_sudo.write(vals)
-
-    def _create_visitor(self, website_page_id=False):
+        is a tracked website page or a tracked view. Only on tracked elements to avoid having
+        too much operations done on every page or other http requests.
+        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked elements."""
+        url = request.httprequest.url
+        website_track_values = {
+            'url': url,
+            'visit_datetime': datetime.now(),
+        }
+        if website_page and website_page.is_tracked:
+            website_track_values['page_id'] = website_page.id
+            domain = [('page_id', '=', website_page.id)]
+        else:
+            domain = [('url', '=', url)]
+        if not visitor_sudo:
+            # If visitor does not exist
+            visitor_sudo = self._create_visitor(website_track_values)
+            sign = visitor_sudo._get_visitor_sign().get(visitor_sudo.id)
+            expiration_date = datetime.now() + timedelta(days=365)
+            response.set_cookie('visitor_id', sign, expires=expiration_date.timestamp())
+        else:
+            visitor_sudo._add_tracking(domain, website_track_values)
+
+    def _add_tracking(self, domain, website_track_values):
+        """ Update the visitor when a website_track is added"""
+        domain = expression.AND([domain, [('visitor_id', '=', self.id)]])
+        last_view = self.env['website.track'].sudo().search(domain, limit=1)
+        if not last_view or last_view.visit_datetime < datetime.now() - timedelta(minutes=30):
+            website_track_values['visitor_id'] = self.id
+            self.env['website.track'].create(website_track_values)
+        self.env['website.visitor.lastconnection'].create({'visitor_id': self.id, 'connection_datetime': website_track_values['visit_datetime']})
```

Comment:
indent 2 lines

Reply:
Don't we want to know if a visitor is online as much as possible ?